crates/wasm-encoder/src/reencode.rs

```diff
@@ -0,0 +1,1577 @@
+    }
+
+    fn const_expr(&mut self, const_expr: wasmparser::ConstExpr) -> Result<crate::ConstExpr> {
+        utils::const_expr(self, const_expr)
+    }
+
+    fn catch(&mut self, arg: wasmparser::Catch) -> crate::Catch {
+        utils::catch(self, arg)
+    }
+
+    fn composite_type(
+        &mut self,
+        composite_ty: wasmparser::CompositeType,
+    ) -> Result<crate::CompositeType> {
+        utils::composite_type(self, composite_ty)
+    }
+
+    fn entity_type(&mut self, type_ref: wasmparser::TypeRef) -> Result<crate::EntityType> {
+        utils::entity_type(self, type_ref)
+    }
+
+    fn export_kind(&mut self, external_kind: wasmparser::ExternalKind) -> crate::ExportKind {
+        utils::export_kind(self, external_kind)
+    }
+
+    fn field_type(&mut self, field_ty: wasmparser::FieldType) -> Result<crate::FieldType> {
+        utils::field_type(self, field_ty)
+    }
+
+    fn func_type(&mut self, func_ty: wasmparser::FuncType) -> Result<crate::FuncType> {
+        utils::func_type(self, func_ty)
+    }
+
+    fn global_type(&mut self, global_ty: wasmparser::GlobalType) -> Result<crate::GlobalType> {
+        utils::global_type(self, global_ty)
+    }
+
+    fn heap_type(&mut self, heap_type: wasmparser::HeapType) -> Result<crate::HeapType> {
+        utils::heap_type(self, heap_type)
+    }
+
+    fn instruction<'a>(&mut self, arg: wasmparser::Operator<'a>) -> Result<crate::Instruction<'a>> {
+        utils::instruction(self, arg)
+    }
+
+    fn memory_type(&mut self, memory_ty: wasmparser::MemoryType) -> crate::MemoryType {
+        utils::memory_type(self, memory_ty)
+    }
+
+    fn mem_arg(&mut self, arg: wasmparser::MemArg) -> crate::MemArg {
+        utils::mem_arg(self, arg)
+    }
+
+    fn ordering(&mut self, arg: wasmparser::Ordering) -> crate::Ordering {
+        utils::ordering(self, arg)
+    }
+
+    fn ref_type(&mut self, ref_type: wasmparser::RefType) -> Result<crate::RefType> {
+        utils::ref_type(self, ref_type)
+    }
+
+    fn storage_type(&mut self, storage_ty: wasmparser::StorageType) -> Result<crate::StorageType> {
+        utils::storage_type(self, storage_ty)
+    }
+
+    fn struct_type(&mut self, struct_ty: wasmparser::StructType) -> Result<crate::StructType> {
+        utils::struct_type(self, struct_ty)
+    }
+
+    fn sub_type(&mut self, sub_ty: wasmparser::SubType) -> Result<crate::SubType> {
+        utils::sub_type(self, sub_ty)
+    }
+
+    fn table_type(&mut self, table_ty: wasmparser::TableType) -> Result<crate::TableType> {
+        utils::table_type(self, table_ty)
+    }
+
+    fn tag_kind(&mut self, kind: wasmparser::TagKind) -> crate::TagKind {
+        utils::tag_kind(self, kind)
+    }
+
+    fn tag_type(&mut self, tag_ty: wasmparser::TagType) -> crate::TagType {
+        utils::tag_type(self, tag_ty)
+    }
+
+    fn val_type(&mut self, val_ty: wasmparser::ValType) -> Result<crate::ValType> {
+        utils::val_type(self, val_ty)
+    }
+
+    fn custom_section<'a>(
```

Comment:
Perhaps both? The current naming I think makes sense but it might also make sense to have a `parse_custom_section` too

Reply:
Thinking toward the future, it might make more sense to have `parse_custom_section` which would parse known custom sections and remap indices in them (e.g. the `name` custom section)

Reply:
Would that have a different signature?

Reply:
In theory yeah, taking a `&mut Module` perhaps instead of returning a `CustomSection`, but `CustomSection` also can own its contents so maybe not. I don't have a strong preference either way.

Reply:
Thanks for the feedback - I've now added an extra parse method that calls the existing method